worker/internal/app/handler.go

```diff
@@ -39,28 +52,114 @@ func (h Handler) DecompressHandler() echo.HandlerFunc {
 
 func (h Handler) WebhookHandler() echo.HandlerFunc {
 	return func(c echo.Context) error {
-		var msg msgBody
-		var w webhook.Webhook
-		if err := c.Bind(&msg); err != nil {
-			if err := c.Bind(&w); err != nil {
-				return err
-			}
-		} else if data, err := msg.Data(); err != nil {
-			return err
-		} else if err := json.Unmarshal(data, &w); err != nil {
+		var webhook webhook.Webhook
+		var err error
+
+		if h.isAWS(c.Request()) {
+			webhook, err = parseSNSWebhookMessage(c.Request().Body)
+		} else if h.isGCP(c.Request()) {
+			webhook, err = parsePubSubWebhookMessage(c, c.Request().Body)
+		} else {
+			err = errors.New("unsupported request source")
+		}
+
+		if err != nil {
+			log.Errorf("failed to parse request body: %s", err.Error())
 			return err
 		}
 
-		if err := h.Controller.WebhookController.Webhook(c.Request().Context(), &w); err != nil {
-			log.Errorf("failed to send webhook. webhook: %#v err:%s", w, err.Error())
+		if err := h.Controller.WebhookController.Webhook(c.Request().Context(), &webhook); err != nil {
+			log.Errorf("failed to send webhook. webhook: %#v err:%s", webhook, err.Error())
 			return err
 		}
 
-		log.Info("webhook has been sent: %#v", w)
+		log.Infof("webhook has been sent: %#v", webhook)
 		return c.NoContent(http.StatusOK)
 	}
 }
 
+func (h Handler) isAWS(r *http.Request) bool {
+	return r.Header.Get("X-Amz-Sns-Message-Type") == "Notification"
+}
+
+func (h Handler) isGCP(r *http.Request) bool {
+	// TODO: need to find a better way to detect GCP requests
+	for headerName := range r.Header {
+		if strings.HasPrefix(headerName, "X-GOOG-") {
+			return true
+		}
+	}
+	return false
+}
+
+func parseSNSDecompressMessage(body io.Reader) (rhttp.DecompressInput, error) {
+	var payload sns.Payload
+	var input rhttp.DecompressInput
+	
+	if err := json.NewDecoder(body).Decode(&payload); err != nil {
+		return input, err
+	}
+	
+	if err := json.Unmarshal([]byte(payload.Message), &input); err != nil {
+		return input, err
+	}
+	
+	// Validates payload's signature
+	if err := payload.VerifyPayload(); err != nil {
+		return input, err
+	}
+	
+	return input, nil
+}
+
+func parsePubSubDecompressMessage(c echo.Context, body io.Reader) (rhttp.DecompressInput, error) {
+	var input rhttp.DecompressInput
+	
+	if err := c.Bind(&input); err != nil {
```

Comment:
Use `json.NewDecoder(body).Decode(&input)` instead of `c.Bind(&input)`

Reply:
I haven't wrote this code.
This is the original code that we have right now for GCP.
Done!